### ISSUE TITLE
Controlling my Computer through Hand Gestures

### ISSUE THREAD

**Issue by Stepheni12** [Stepheni12/iSteph#2]
---
title: Controlling my Computer through Hand Gestures
description: I developed a program that allows you to control your browser with finger movements detected by your webcam.
slug: gesture-computing
tags: google mediapipe, pyautogui, opencv, computer vision 
category: blog
date: 2024-02-16
---

The whole AI space has been flooded with everything and anything related to LLMs. Lately, I have been spending a lot of time working on projects solely in that space, which has been exciting, but the other day I decided to change things up with a side project. The general idea is being able to control your computer with finger movements. I was inspired to do this by something I came across online a while ago, and I know my younger self would appreciate a cool project like this.

TL;DR Control Google Chrome with your fingers

https://github.com/Stepheni12/iSteph/assets/26886046/8f62cbc5-ce5e-4b9d-b382-fa57136f0085

### Introduction
You can go so many directions with this, as you are essentially just detecting finger movements and then running different sections of code depending on the movement. I'll get into some other potential project directions at the end. For this project, I decided to link finger movements to different useful browser commands so you can navigate around your browser using only your hand. To do this, you'll need a couple of useful Python libraries. Using these libraries will handle 90% of the workload.

- [MediaPipe](https://developers.google.com/mediapipe) - Python library developed by Google with a suite of models available for on-device ML.
- [PyAutoGUI](https://pyautogui.readthedocs.io/en/latest/) - Python library for programmatically controlling mouse and keyboard events.
- [OpenCV](https://opencv.org/get-started/) - Python computer vision library with a lot of functionality.

### Understanding MediaPipe
MediaPipe has a variety of ML models available across vision, NLP, and audio. For this project, I utilized the  Gesture Recognition model which has two versions: one detects specific gestures from a list of known gestures, and the other detects hand landmarks and outputs their coordinate location. This photo shows the hand landmarks that it labels.

<img width="1073" alt="hand-landmarks" src="https://github.com/Stepheni12/iSteph/assets/26886046/79d74f2b-d820-4c20-b1c5-d650a7982ce2">

I used the raw hand landmarks version and then detected the gestures in post-processing. If you've got the dataset, you can train the model to learn new gestures, but the hand landmarks were optimal for my task.

### Image Processing with OpenCV
To port the webcam feed into the model to obtain the hand landmarks, I used OpenCV for all the image processing. This library makes it easy to accomplish. You define a VideoCapture object, the webcam, and then the program repeatedly extracts frames and processes them. My nearly 10-year-old laptop had no issues capturing the webcam feed at 30 FPS and running the model against the input frames. 

MediaPipe also includes functionality to visualize the hand landmarks as they are processed. That visualization code, in combination with OpenCV, allows you to get an idea of what the model is seeing and making decisions based on!

![hand_pic](https://github.com/Stepheni12/iSteph/assets/26886046/d3d8420b-9296-49b2-8b93-117f68f4899d)

### Detecting Gestures
In my system, I implemented distinct finger movements to regulate browser functions, requiring the individual bending of each finger. This design resulted in five distinct controls, each dedicated to performing specific actions.

- Thumb: Close tab
- Index: Scroll up
- Middle: Scroll down
- Ring: Switch tabs
- Pinky: Open new tab

Detecting these movements was straightforward. If you reference the hand landmark image with the numbered labels above, you can see that every finger has a landmark at the tip and the base. For example, the index fingertip is 8, and the base is 5. To detect when the index finger bends, I compare the current distance of landmarks 8 and 5 with the baseline distance. I then just repeated this process for every finger with their respective landmarks. The formula for measuring the bent finger is the same for every finger except the thumb. I had to lower the threshold on that a bit to make it easier to detect.

### PyAutoGUI for Keyboard Events
Using the gesture detection method above, I linked the detection of the different gestures to different PyAutoGUI keyboard commands.

- Thumb: pyautogui.hotkey('ctrl', 'w')
- Index: pyautogui.scroll(800) `This number is relative to your screen size so adjust accordingly`
- Middle: pyautogui.scroll(-800)
- Ring: pyautogui.hotkey('ctrl', 'tab')
- Pinky: pyautogui.hotkey('ctrl', 't')

### Complete System
The flow of the overall system once you run the program is as follows:
1. PyAutoGUI activates your Google Chrome window and brings it into focus.
2. Let the user know to place their hand in front of the camera for 3 seconds to calibrate.
3. Once calibrated, you can begin finger movements. There is a two-second delay between detecting gestures.
> Currently, there's no good way to pull your hand off the screen if you don't want to input a gesture without it incorrectly triggering a random finger's action. The best way to remove your hand from the camera is to make a gesture and then pull your hand from view during the delay between detections as this will ensure that an unintended action does not occur.
4. Once your hand is off-screen and you want to make another gesture, you need to hold it in front of the camera, as done in the pictures above, for about three seconds before it will start detecting gestures again.

### Outcome
When you know how to use it, the results are great. From a practical point of view, this could be useful if all the commands were for browsing or reading. Perhaps instead of a new tab command, you could replace that with changing the window so now you can jump from scrolling on one window to scrolling on another. Imagine yourself kicked back in your reclined office chair using your fingers to browse various reading materials on your computer without having to unrecline (definitely not a word) to get back to your mouse and keyboard!

#### Removing your hand from the screen
In the current state, if you have your hand up for at least 3 seconds the system expects a gesture, but if you don't perform one and you pull your hand out of camera view, it will just perform any one of the commands depending on how it weirdly measured the landmarks as your hand left the view. It isn't an issue if you know how the system works because the way to handle this correctly is to make a gesture and then pull your hand from view during the delay after the keyboard event takes place. Similarly, when you bring your hand back into view, you need to hold it there for 3 seconds until it looks to detect a gesture. This delay is in place so that as it makes weird measurements when your hand is entering back into view, it doesn't perform random commands.

#### Only looks for Chrome
It's more of a preference, but this version only looks to bring a Chrome window to focus.

#### It doesn't intelligently select the Chrome window
If you have multiple Chrome windows (windows, not tabs) open it will select whichever PyAutoGUI first finds and bring that one to focus when the program begins.

#### Potential Directions
A list of different directions you can go with this general baseline project.
- The MediaPipe Library can detect multiple hands, so maybe develop a system that incorporates both hands.
- You can link the gestures to any other thing you can program, so the possibilities are endless.
    - Opening a Google Chrome tab ready to compose an email.
    - Sound effects so you use your hand as a soundboard.
    - Smart home appliances so your fingers can operate them (i.e. turning on lights).
    - PyAutoGUI to control the mouse and keyboard and do just about anything.
    - OpenAI API calls.
- Create a UI that makes it easier to use.
- Implement a more robust system that better handles the nuances.
- Maybe instead of hand gestures, you create a system that detects certain face gestures
- Further train the MediaPipe model to teach it different gestures.
